gnovm/cmd/gno/transpile.go

```diff
@@ -259,8 +292,116 @@ func goBuildFileOrPkg(fileOrPkg string, cfg *transpileCfg) error {
 	goBinary := cfg.goBinary
 
 	if verbose {
-		fmt.Fprintf(os.Stderr, "%s\n", fileOrPkg)
+		fmt.Fprintf(os.Stderr, "%s [build]\n", filepath.Clean(fileOrPkg))
+	}
+
+	return buildTranspiledPackage(fileOrPkg, goBinary)
+}
+
+// getPathsFromImportSpec returns the directory paths where the code for each
+// importSpec is stored (assuming they start with [transpiler.ImportPrefix]).
+func getPathsFromImportSpec(rootDir string, importSpec []*ast.ImportSpec) (dirs []string) {
+	for _, i := range importSpec {
+		path, err := strconv.Unquote(i.Path.Value)
+		if err != nil {
+			continue
+		}
+		if strings.HasPrefix(path, transpiler.ImportPrefix) {
+			res := strings.TrimPrefix(path, transpiler.ImportPrefix)
+
+			dirs = append(dirs, rootDir+filepath.FromSlash(res))
+		}
+	}
+	return
+}
+
+// buildTranspiledPackage tries to run `go build` against the transpiled .go files.
+//
+// This method is the most efficient to detect errors but requires that
+// all the import are valid and available.
+func buildTranspiledPackage(fileOrPkg, goBinary string) error {
+	// TODO: use cmd/compile instead of exec?
+	// TODO: find the nearest go.mod file, chdir in the same folder, trim prefix?
+	// TODO: temporarily create an in-memory go.mod or disable go modules for gno?
+	// TODO: ignore .go files that were not generated from gno?
+
+	info, err := os.Stat(fileOrPkg)
+	if err != nil {
+		return fmt.Errorf("invalid file or package path %s: %w", fileOrPkg, err)
+	}
+	var (
+		target string
+		chdir  string
+	)
+	if !info.IsDir() {
+		dstFilename, _ := transpiler.TranspiledFilenameAndTags(fileOrPkg)
+		// Makes clear to go compiler that this is a relative path,
+		// rather than a path to a package/module.
+		// can't use filepath.Join as it cleans its results.
+		target = filepath.Dir(fileOrPkg) + string(filepath.Separator) + dstFilename
+	} else {
+		// Go does not allow building packages using absolute paths, and requires
+		// relative paths to always be prefixed with `./` (because the argument
+		// go expects are import paths, not directories).
+		// To circumvent this, we use the -C flag to chdir into the right
+		// directory, then run `go build .`
+		chdir = fileOrPkg
+		target = "."
+	}
+
+	// pre-alloc max 5 args
+	args := append(make([]string, 0, 5), "build")
+	if chdir != "" {
+		args = append(args, "-C", chdir)
+	}
+	args = append(args, "-tags=gno", target)
+	cmd := exec.Command(goBinary, args...)
+	out, err := cmd.CombinedOutput()
+	if _, ok := err.(*exec.ExitError); ok {
```

Comment:
Why not check with `errors.Is`?

Reply:
it should be errors.As here, but fair point